src/Prefetch.cpp

```diff
@@ -0,0 +1,228 @@
+#include <algorithm>
+#include <map>
+#include <string>
+
+#include "Prefetch.h"
+#include "IRMutator.h"
+#include "Bounds.h"
+#include "Scope.h"
+#include "Util.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::string;
+using std::vector;
+
+namespace {
+
+// We need to be able to make loads from a buffer that refer to the
+// same original image/param/etc. This visitor finds a load to the
+// buffer we want to load from, and generates a similar load, but with
+// different args.
+class MakeSimilarLoad : public IRVisitor {
+public:
+    const std::string &buf_name;
```

Comment:
This should just be const string, not constref-string

also, nit: we should use the "std::" prefix (or not) consistently within this file, but we're not